### PR TITLE
Support augment_platform in BB

### PR DESCRIPTION
**Pull request by vchuravy**
Companion to https://github.com/JuliaPackaging/JLLWrappers.jl/pull/35

### CODE REVIEW

File: src/AutoBuild.jl
Comment (staticfloat) [inline src/AutoBuild.jl:1277]: I'd rather we write this out to `.pkg/augment_platform.jl`, and then conditionally `include()` it.  And can you also write out a `select_artifacts.jl` file that is [like this one] (https://github.com/JuliaLang/Pkg.jl/blob/2dd335da3dddf492ea03d765061ca7fd6e6c65c2/test/test_packages/AugmentedPlatform/.pkg/select_artifacts.jl).  That will future-proof us for `Pkg.add()`-time detection.

File: src/AutoBuild.jl
Comment (staticfloat) [inline src/AutoBuild.jl:1281]: Since we already have the `@isdefined()` in JLLWrappers, is this necessary?

**Review by staticfloat — CHANGES_REQUESTED** [review]

File: src/AutoBuild.jl
Comment (giordano) [inline src/AutoBuild.jl:716]: I'm doing this in a separate PR (#1129), where I'm also going to add tests for that

Comment (giordano) [reply src/AutoBuild.jl]: #1129 merged and this PR rebased on master

File: src/AutoBuild.jl
Comment (giordano) [inline src/AutoBuild.jl:153]:
```suggestion
    and may be removed in a future version of this package. Please use them sparingly.
```

File: Manifest.toml
Comment (giordano) [inline Manifest.toml:144]: I think we should merge https://github.com/JuliaPackaging/JLLWrappers.jl/pull/35 before actually merging this PR, and use here a released version.

File: src/AutoBuild.jl
Comment (giordano) [inline src/AutoBuild.jl:143]: Should we make this more general?  It isn't really specific to platform-augmentation, especially following https://github.com/JuliaPackaging/BinaryBuilder.jl/pull/1128#discussion_r770987505.  I can see this argument can be (ab)used for other purposes as well.  What about `top_level_block` or something along this line?

File: src/AutoBuild.jl
Comment (vchuravy) [reply src/AutoBuild.jl:1278]: What would be the include condition?

Comment (staticfloat) [reply src/AutoBuild.jl:1278]: `if isfile()`?

File: Project.toml
Comment (giordano) [inline Project.toml:14]: This isn't actually used in the package, as far as I can see
```suggestion
```

File: src/AutoBuild.jl
Comment (giordano) [reply src/AutoBuild.jl:143]: Scratch that, with new changes this is very specialised to platform-augmentation

File: test/jll.jl
Comment (giordano) [inline test/jll.jl:27]:
```suggestion
    @test project["compat"] == Dict("julia" => "1.0", "XZ_jll" => "=2.4.6", "JLLWrappers" => "1.2.0")
```

File: src/AutoBuild.jl
Comment (giordano) [inline src/AutoBuild.jl:1354]: Just to understand, this is needed because in a bare module `include` is not defined?  A comment about this may help future readers :slightly_smiling_face: 

Do we need to define this function (is `include` used elsewhere?) or can just do
```suggestion
            Base.include(@__MODULE__, joinpath("..", ".pkg", "platform_augmentation.jl"))
```
?

**Review by giordano — CHANGES_REQUESTED** [review]

File: src/AutoBuild.jl
Comment (vchuravy) [reply src/AutoBuild.jl]: Yeah!

**Comment by maleadt** [comment]
Rebased and addressed review comments. I'd like to use this for https://github.com/JuliaPackaging/Yggdrasil/pull/4292, which in turn will unblock some work on LLVM.jl (handling assertion builds).

**Comment by maleadt** [comment]
https://github.com/JuliaPackaging/Yggdrasil/pull/4292 has finished CI, so this looks good to go (unless there's further review comments, of course).

**Comment by vchuravy** [comment]
Thanks Tim!

File: src/AutoBuild.jl
Comment (giordano) [inline src/AutoBuild.jl:552]: Not sure, should this use `DEFAULT_JLLWRAPPERS_VERSION_SPEC`?
```suggestion
    jllwrappers_compat = isempty(augment_platform_block) ? DEFAULT_JLLWRAPPERS_VERSION_SPEC : "1.4.0"
```

**Review by giordano — APPROVED** [review]
Looks good to me, thanks a lot @vchuravy @maleadt!

**Pull request merged**
